File: src/shader/conv2d/params.c
// conv2d的input数据

// 常量
// 卷积核
const int length_shape_filter = LENGTH_SHAPE_FILTER;
const int width_shape_filter = WIDTH_SHAPE_FILTER;
const int height_shape_filter = HEIGHT_SHAPE_FILTER;
const int width_texture_filter = WIDTH_TEXTURE_FILTER;
const int height_texture_filter = HEIGHT_TEXTURE_FILTER;
// todo:新增
const int channel_filter = CHANNEL_FILTER;

// 输入数据
const int width_shape_origin = WIDTH_SHAPE_ORIGIN;
const int height_shape_origin = HEIGHT_SHAPE_ORIGIN;
// todo: 新增
const int length_shape_origin = LENGTH_SHAPE_ORIGIN;
const int width_texture_origin = WIDTH_TEXTURE_ORIGIN;
const int height_texture_origin = HEIGHT_TEXTURE_ORIGIN;
const int channel_origin = CHANNEL_ORIGIN;

// 输出数据
const int width_shape_out = WIDTH_SHAPE_OUT;
const int height_shape_out = HEIGHT_SHAPE_OUT;
// todo: 新增
const int width_texture_out = WIDTH_TEXTURE_OUT;
const int height_texture_out = HEIGHT_TEXTURE_OUT;
const int channel_out = CHANNEL_OUT;

// 计算相关
// 拆分步长
const int stride_h = STRIDE_HORIZONTAL;
const int stride_v = STRIDE_VERTICAL;
// padding的数目
const int padLeft = PAD_LEFT;
const int padTop = PAD_TOP;
// dilation膨胀系数
const int dilation_h = DILATION_HORIZONTAL;
const int dilation_v = DILATION_VERTICAL;


// uniform变量
// 卷积核
uniform int numbers_shape_filter[LENGTH_SHAPE_FILTER];
uniform int numbers_shape_origin[LENGTH_SHAPE_ORIGIN];
uniform sampler2D texture_filter;

// 输入数据
uniform sampler2D texture_origin;


// varying变量
// 顶点shader透传的材质坐标
varying vec2 vCoord;

File: packages/paddlejs-backend-webgpu/src/ops/conv2d/params.ts
/**
 * @file conv2d params
 */

export default function(params) {
    const {
        width_shape_filter: widthShapeFilter,
        height_shape_filter: heightShapeFilter,
        channel_filter: channelFilter,
        total_shape_filter: totalShapeFilter,
        total_shape_origin: totalShapeOrigin,
        width_shape_origin: widthShapeOrigin,
        height_shape_origin: heightShapeOrigin,
        channel_origin: channelOrigin,
        width_shape_out: widthShapeOut,
        height_shape_out: heightShapeOut,
        width_texture_out: widthTextureOut,
        channel_out: channelOut,
        channel_bias: channelBias,
        width_shape_bias: widthShapeBias,
        height_shape_bias: heightShapeBias,
        total_shape_bias: totalShapeBias,
        strides,
        paddings,
        dilations,
        groups,
        binding_origin: bindingOrigin,
        binding_filter: bindingFilter,
        binding_out: bindingOut,
        binding_bias: bindingBias
    } = params;

    return `

    // 卷积核
    const int width_shape_filter = ${widthShapeFilter};
    const int height_shape_filter = ${heightShapeFilter};
    const int channel_filter = ${channelFilter};
    const int total_shape_filter = ${totalShapeFilter};

    // origin
    const int total_shape_origin = ${totalShapeOrigin};
    const int width_shape_origin = ${widthShapeOrigin};
    const int height_shape_origin = ${heightShapeOrigin};
    const int channel_origin = ${channelOrigin};

    // bias
    const int channel_bias = ${channelBias};
    const int width_shape_bias = ${widthShapeBias};
    const int height_shape_bias = ${heightShapeBias};
    const int total_shape_bias = ${totalShapeBias};
    // output
    const int width_shape_out = ${widthShapeOut};
    const int height_shape_out = ${heightShapeOut};
    const int width_texture_out = ${widthTextureOut};
    const int channel_out = ${channelOut};

    // 计算相关
    // 拆分步长
    const int stride_h = ${strides[0]};
    const int stride_v = ${strides[1]};
    // padding的数目
    const int padLeft = ${paddings[0]};
    const int padTop = ${paddings[1]};
    // dilation膨胀系数
    const int dilation_h = ${dilations[0]};
    const int dilation_v = ${dilations[1]};
    // groups
    const int groups = ${groups};

    layout(std430, set = 0, binding = ${bindingBias}) readonly buffer BiasMatrix {
        float numbers[];
    } biasMatrix;

    layout(std430, set = 0, binding = ${bindingOrigin}) readonly buffer Origin {
        float numbers[];
    } originMatrix;

    layout(std430, set = 0, binding = ${bindingFilter}) readonly buffer Filter {
        float numbers[];
    } filterMatrix;

    layout(std430, set = 0, binding = ${bindingOut}) buffer Out {
        float numbers[];
    } outMatrix;

    `;
}
